### ISSUE TITLE
Javadoc: external links to classes in java.lang are not resolved

### ISSUE THREAD

**Issue by Stewori** [doxygen/doxygen#8541]
**Describe the bug**
Assume a tagfile of the module ``java.base`` is located at ``/path_to_doxygen_output_of_java.base_1.14/java.base_1.14_tagfile.xml`` and declared in the Doxygen config file under ``TAGFILES``.
Processing Java code
```
import java.util.Vector;

public class Test {
    public String a;
    public Vector b;
}
```
results in the following xml:
```
...
<compounddef id="classTest" kind="class" language="Java" prot="public">
    <compoundname>Test</compoundname>
      <sectiondef kind="public-attrib">
      <memberdef kind="variable" id="classTest_1a4dfe3f6d6cbfd2814da6d802e9d916c3" prot="public" static="no" mutable="no">
        <type>String</type>
        <definition>String Test.a</definition>
        <argsstring></argsstring>
        <name>a</name>
        <briefdescription>
        </briefdescription>
        <detaileddescription>
        </detaileddescription>
        <inbodydescription>
        </inbodydescription>
        <location file="Test.java" line="4" column="19" bodyfile="Test.java" bodystart="4" bodyend="-1"/>
      </memberdef>
      <memberdef kind="variable" id="classTest_1a184487560da0c075f32b0cbd2261afe2" prot="public" static="no" mutable="no">
        <type><ref refid="classjava_1_1util_1_1Vector" kindref="compound" external="/path_to_doxygen_output_of_java.base_1.14/java.base_1.14_tagfile.xml">Vector</ref></type>
        <definition>Vector Test.b</definition>
        <argsstring></argsstring>
        <name>b</name>
        <briefdescription>
        </briefdescription>
        <detaileddescription>
        </detaileddescription>
        <inbodydescription>
        </inbodydescription>
        <location file="Test.java" line="5" column="19" bodyfile="Test.java" bodystart="5" bodyend="-1"/>
      </memberdef>
      </sectiondef>
...
```
Notice how ``Vector`` is properly linked to its external resource while ``String`` is not. This is because Doxygen does not consider that the package ``java.lang`` is always implicitly imported. The same issue occurs without external linking if one processes the source of the ``java.base`` module with Doxygen. No internal links to classes in ``java.lang`` are created although there should be plenty of them.

**Expected behavior**
I expect the xml output to be the same as if one had prepended the line ``import java.lang.*`` to every ``.java`` file.
The following expected output can be produced by running Doxygen on the example above with an explicit import of ``java.lang`` prepended, i.e.
```
import java.lang.*;
import java.util.Vector;

public class Test {
    public String a;
    public Vector b;
}
```
results in
```
...
<compounddef id="classTest" kind="class" language="Java" prot="public">
    <compoundname>Test</compoundname>
      <sectiondef kind="public-attrib">
      <memberdef kind="variable" id="classTest_1a4dfe3f6d6cbfd2814da6d802e9d916c3" prot="public" static="no" mutable="no">
        <type><ref refid="classjava_1_1lang_1_1String" kindref="compound" external="/path_to_doxygen_output_of_java.base_1.14/java.base_1.14_tagfile.xml">String</ref></type>
        <definition>String Test.a</definition>
        <argsstring></argsstring>
        <name>a</name>
        <briefdescription>
        </briefdescription>
        <detaileddescription>
        </detaileddescription>
        <inbodydescription>
        </inbodydescription>
        <location file="Test.java" line="5" column="19" bodyfile="Test.java" bodystart="5" bodyend="-1"/>
      </memberdef>
      <memberdef kind="variable" id="classTest_1a184487560da0c075f32b0cbd2261afe2" prot="public" static="no" mutable="no">
        <type><ref refid="classjava_1_1util_1_1Vector" kindref="compound" external="/path_to_doxygen_output_of_java.base_1.14/java.base_1.14_tagfile.xml">Vector</ref></type>
        <definition>Vector Test.b</definition>
        <argsstring></argsstring>
        <name>b</name>
        <briefdescription>
        </briefdescription>
        <detaileddescription>
        </detaileddescription>
        <inbodydescription>
        </inbodydescription>
        <location file="Test.java" line="6" column="19" bodyfile="Test.java" bodystart="6" bodyend="-1"/>
      </memberdef>
      </sectiondef>
...
```

**To Reproduce**
Find a Java file and doxygen config attached. The issue is demonstrated like mentioned above.
After running doxygen on the example, the files in the subfolder ``xml`` contain the snippet shown above.
[issue_external_linking_java.lang.zip](https://github.com/doxygen/doxygen/files/6451397/issue_external_linking_java.lang.zip)

If requested I can provide instructions on how to build a Doxygen tagfile for java.base. I would also provide the tagfile but it is >21MB and sharing it would probably require adding all the license docs required by the GPL.

**Version**
Tested with doxygen clone from 2021-04-11:
```
/blah/doxygen/build/bin/doxygen -v
1.9.2 (e9185650216f3c3fa97759caf67ee79db66cb5e1)
```

`lsb_release -a`:
```
LSB Version:	core-9.20160110ubuntu0.2-amd64:core-9.20160110ubuntu0.2-noarch:security-9.20160110ubuntu0.2-amd64:security-9.20160110ubuntu0.2-noarch
Distributor ID:	LinuxMint
Description:	Linux Mint 18 Sarah
Release:	18
Codename:	sarah
```

**Additional context**
Linking to a tagfile of java.base is an essential part to work around issue #8433. Especially the classes from ``java.lang`` are omnipresent in most projects.

**Comment by Stewori** [comment]
Sorry that was just an edit fail due to unlucky drag and drop. For simplicity, I will edit the initial post to correct that issue.

**Comment by albert-github** [comment]
I now only see the difference:
```
<         <type>String</type>
---
>         <type><ref refid="classjava_1_1lang_1_1String" kindref="compound" external="/path_to_doxygen_output_of_java.base_1.14/java.base_1.14_tagfile.xml">String</ref></type>
```
is this the difference you meant? 

Besides the 2 questions posed in my previous comment (https://github.com/doxygen/doxygen/issues/8541#issuecomment-836557762) a 3rd question:
- what is the size of the tag file when compressed?

**Comment by Stewori** [comment]
> is this the difference you meant?

Yes, exactly.

> what is the size of the tag file when compressed

The tagfile is just 2.4MB when zip-compressed (1.9MB with bz2). To compress it should have occurred to me in first place. I guess it's fine to add it here since it is not derived from an input file but merely a list of files and classes. And then there also is the fair use principle, so here it is:
[java.base_1.14_tagfile.xml.zip](https://github.com/doxygen/doxygen/files/6455430/java.base_1.14_tagfile.xml.zip)

> does it make any difference when you use the syntax TAGFILES = file1=loc1 as at the moment you use TAGFILES = file1?

No, that does not appear to make a difference. This was to be expected, because

- one can manually workaround the issue for a particular file by prepending ``import java.lang.*;`` regardless of the syntax used in ``TAGFILES``.
- the issue also occurs internally in ``java.base`` without assigning a value to ``TAGFILES`` at all

**Comment by albert-github** [comment]
I think the observation:
> This is because Doxygen does not consider that the package java.lang is always implicitly imported.

is probably the right observation, but is this somewhere officially documented / standardized for Java or is this a de facto standard?

**Comment by albert-github** [comment]
I think I can answer this myself from the "The Java® Language, Specification, Java SE 16 Edition" ,Chapter 7, Packages and Modules:

> Code in a compilation unit automatically has access to all classes and interfaces
declared in its package and also automatically imports all of the public classes and
interfaces declared in the predefined package java.lang.

So it is officially documented.

**Comment by Stewori** [comment]
I confirm that your proposed patch fixes the issue. Thanks for fixing it so quickly!

**Comment by albert-github** [comment]
Code has been integrated in master on github (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.2.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
